blog/content/posts/2025-12-xx-transport-direction.md

```diff
@@ -0,0 +1,83 @@
+
+### A Stateless Protocol
+
+At the outset, MCP was designed to be _stateful_, with Clients and Servers maintaining knowlege of each other over a bidirectional communication channel.
+
+Connections are initialized with a handshake which shares information like Capabilities and Protocol Version. This state remains fixed for the duration of the connection, and requires techniques such as sticky sessions or distributed session storage to for scaled deployment.
+
+We propose making MCP stateless by:
+
+- Replacing the `initialize` handshake - and sending the shared information with each request/response instead.
+- Providing a `discovery` mechanism for Clients to query Server Capabilities if they need the information early (for example for UX purposes).
+
+We're moving toward a more dynamic model where clients can optimistically "just try" what they need and receive helpful error messages if unsupported.
+
+> [!NOTE]
+>
+> Many SDKs present a _`stateless`_ option in their Server transport configuration. This flag actually controls whether the `Mcp-Session-Id` header is used - read below for more on sessions.
+
+### Elevating Sessions
+
+Today, Sessions are a side effect of a transport connection. For STDIO, sessions are implicit in the process lifecycle. For Streamable HTTP, sessions are created when a Server assigns an `Mcp-Session-Id` during `intialize`. This mixes transport and application layer concerns.
+
+We plan to move Sessions to the _data model layer_ - making them explicit rather than implicit.
+
+This means that MCP applications will be able to handle sessions as part of their domain logic. A cookie style mechanism is the preferred choice.
+
+This makes MCP similar to standard HTTP - where the protocol itself is stateless, while applications build stateful semantics with cookies, tokens and similar mechanisms. The exact approach to session creation is still being worked on - but this removes existing ambiguities on what a session is in remote MCP.
+
+### Elicitations and Sampling
+
+Two standout MCP features enable advanced AI workflows: [Elicitations](https://spec.modelcontextprotocol.io/specification/2025-03-26/client/elicitation) for requesting human input, and [Sampling](https://spec.modelcontextprotocol.io/specification/2025-03-26/client/sampling) for agentic LLM interactions.
+
+In our stateless design, this bidirectional request pattern requires rethinking. Currently when a Server needs more information to complete a Tool Call, it suspends operation and waits for a Client response - meaning it must remember all of its outstanding requests.
+
+To avoid the need to do this, we'll make the Server Request/Response similar to the way Chat APIs work. This means that the Client will return both the Request _and_ Response allowing the Server to reconstruct the necessary state purely from the returned message.
```

Comment:
```suggestion
To avoid this problem, we'll make the server request and response similar to the way chat APIs work. The client will return both the request _and_ response, allowing the server to reconstruct the necessary state purely from the returned message.
```